### PR TITLE
feat(scheduler): Handled errors with specific method

### PR DESCRIPTION
**Pull request by eddiebergman**
This PR enables the scheduler to treat different errors in different ways, for example treating a `"raise"` when a `ValueError` happens but `"ignore"` a `TypeError`.

```python
scheduler.run(on_exception={ValueError: "raise", TypeError: "ignore"})
```

The default behaviour when an exception is not a subclass of those provided is to `"raise"`.
The full behaviour is best explained by the docs and the tests in the PR.

Users should not be allowed to explicitly pass `BaseException` or `KeyboardInterupt` (a subclass of `BaseException`) as these usually require shutting down fast and should almost never be allowed to be ignored.

The previous work-around to this problem would be to do the following, where it would not be clear what to do:

```python
@task.on_exception
def handle_exception(_, e: BaseException) -> None:
    if isinstance(e, ExceptionIKnowAbout):
		scheduler.stop(msg="Where do I get this from?", exception=e)
	elif isinstance(e, SomeOtherException):
		scheduler.stop(msg="This list could get long...", exception=e)
	else:
		raise e

scheduler.run(on_exception="ignore")
```

The above work-around still does not handle some other side cases which are now more explicitly handled.

---

The reason for this PR is best illustrated with an example. In the below code, we evaluate a machine learning model which we may expect to raise a memory error. During development, a `ValueError` occured, which we would like to still be raised.

In the example below, the `Pynisher.MemoryError` raises sometimes... but not all the time.

```python
def eval(trial: Trial) -> Trial.Report:
	# ...
	with trial.begin():
		model.fit()  # Pynisher.MemoryError here

    raise ValueError("oops")  # Some user bug happened while developing

	if trial.exception:
		return trial.fail(...)

	return trial.success(...)

task = Task(eval, plugins=PynisherPlugin(memory_limit=(1, "mb")))
```

A `MemoryError` may occur and kill the whole task before any `Trial.Report` can be returned. If the user is aware of this, they may wish to `"ignore"` this in the `Scheduler`, opting to handle it manually with `@task.on_exception`. However a blanket `"ignore"` would also prevent the `ValueError` from causing the `Scheduler` to stop, perhaps not informing the user of a silent bug they may wish to not ignore during development

Previously there was no way to distinguish and handle these exceptions separately, with both being treated equally. Now the solution is:

```python
# The below would ignore the memory error, allowing users to handle it with 
scheduler.run(on_exception={PynisherPlugin.PynisherException: "ignore"})

# `@task.on_exception`, while still allowing them to develop and catch the ValueError
@task.on_exception
def handle_exception(...):
    # Do something about the memory error
```

### CODE REVIEW

**Comment by eddiebergman** [comment]
Side question: Would it be too weird to even allow providing functions instead of `"raise" | "ignore" | "end"` that are deffered to for error handling? Afraid it might be a bit too dynamic but that would give the most flexibility from what I can gather.

**Comment by LennartPurucker** [comment]
> Side question: Would it be too weird to even allow providing functions instead of `"raise" | "ignore" | "end"` that are deffered to for error handling? Afraid it might be a bit too dynamic but that would give the most flexibility from what I can gather.

It's too specific for a real use case, in my opinion. Also, the overhead is not worth it as long as we do not know of a use case for this.

**Review by LennartPurucker — APPROVED** [review]
Hard to give a good review on this without more time but otherwise LGTM

**Comment by eddiebergman** [comment]
> Hard to give a good review on this without more time but otherwise LGTM

Absolutely fair, the only thing you can really review is the tests, if you wouldn't mind

File: src/amltk/scheduling/scheduler.py [hunk 1]
```diff
@@ -1256,6 +1307,38 @@ def run(
...
+
+                    scheduler.run(
+                        on_exception={
+                            PynisherPlugin.Exception: "ignore",
+                            Exception: "raise"  # Default...
+                        }
+                    )
+                    ```
+
+                    !!! tip "Order"
```
Comment (LennartPurucker) [inline src/amltk/scheduling/scheduler.py:1330]: IMO, test the order thing as well.

**Review by LennartPurucker — APPROVED** [review]
LGTM the tests

**Pull request merged**